clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -157,15 +157,23 @@ public void onMetadataUpdate(Cluster cluster) {
                 if (!cluster.unauthorizedTopics().isEmpty())
                     throw new TopicAuthorizationException(new HashSet<>(cluster.unauthorizedTopics()));
 
-                if (subscriptions.hasPatternSubscription()) {
+                boolean hasAutoTopicSubscription = subscriptions.hasAutoTopicSubscription();
+                boolean hasPatternSubscription = subscriptions.hasPatternSubscription();
+
+                if (hasAutoTopicSubscription || hasPatternSubscription) {
                     final Set<String> topicsToSubscribe = new HashSet<>();
+                    Set<String> internalTopics = cluster.internalTopics();
+                    Set<String> topics = hasAutoTopicSubscription ? subscriptions.subscription() : cluster.topics();
 
-                    for (String topic : cluster.topics())
-                        if (subscriptions.getSubscribedPattern().matcher(topic).matches() &&
-                                !(excludeInternalTopics && cluster.internalTopics().contains(topic)))
+                    for (String topic: topics)
+                        if (!(excludeInternalTopics && internalTopics.contains(topic)) &&
+                                (hasAutoTopicSubscription || subscriptions.getSubscribedPattern().matcher(topic).matches()))
                             topicsToSubscribe.add(topic);
 
-                    subscriptions.subscribeFromPattern(topicsToSubscribe);
+                    if (hasAutoTopicSubscription)
```

Comment:
@vahidhashemian Good question. For assign(), it makes sense to call commitAsync() prior to changing the assignment (arguably, it should be commitSync() but that would make assign() a blocking call). I think it also makes sense for unsubscribe since the current behavior is to remove the assignment immediately when it is called. Actually that seems reasonable since we also immediately leave the group (when using dynamic assignment).